src/v/archival/ntp_archiver_service.cc

```diff
@@ -1696,6 +1696,27 @@ ss::future<ntp_archiver::manifest_updated> ntp_archiver::garbage_collect() {
         co_return updated;
     }
 
+    // If we are about to delete segments, we must ensure that the remote
```

Comment:
This ends up being a bit chatty because we _also_ then upload the manifest after gc in housekeeping().  I'm somewhat tempted to remove that second upload call, because we don't care much about the update to the `replaced` list: it's this earlier point where we are about to delete objects that it's critical to reflect in the remote manifest state.

Thoughts @VladLazar ?

Reply:
Yeah, conceptually the desired steps go from:

```
retention_updated_manifest = apply_retention() # step 1: move segments into replaced list
gc_updated_manifest = garbage_collect() # step 3: remove segments, clear replaced list, may upload manifest
if (any manifest update):
  upload and flush manifest # step 2: upload manifest
```

to

```
retention_updated_manifest = apply_retention() # step 1: move segments into replaced list
if (retention_updated_manifest):
  upload and flush manifest # step 2: upload manifest (maybe should be lumped into apply_retention() or garbage_collect()?)
garbage_collect() # step 3: remove segments, clear replaced list, _leave the manifest dirty_
```

It's probably fine and generally safer to lump the upload into the `garbage_collect()` step as we have in this PR, since we can already check `retention_updated_manifest` by seeing whether the manifest has anything to remove.

We just need to be cognizant that `cleanup_metadata()` may be applied multiple times, e.g. in case a follower downloads the manifest with an empty replaced list, and then the leader tells it to clean a replaced list that it doesn't have (which seems fine? it looks idempotent)

Reply:
The change makes sense to me. I'd also be tempted to drop the post-housekeeping upload (or make it conditional). Clusters that are the source for read-replicas could just use smaller manifest upload interval. That makes the semantics clearer IMO: manifest will be uploaded every upload period and that's when the updates become visible to read replicas.